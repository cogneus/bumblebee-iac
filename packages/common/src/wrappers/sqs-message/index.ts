export * from './sqs-message.wrapper'
export * from './types/error-action-collection.type'
