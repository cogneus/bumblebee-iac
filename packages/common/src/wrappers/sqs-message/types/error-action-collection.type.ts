export type ErrorAction = 'no-retry' | 'ignore'
export type ErrorActionCollection = { [key in string]?: ErrorAction }
