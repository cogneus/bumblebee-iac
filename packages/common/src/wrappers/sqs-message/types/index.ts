export * from './error-action-collection.type'
export * from './sqs-message-dependencies.interface'
export * from './sqs-message-handler-response.interface'
export * from './sqs-message-parser.type'
export * from './sqs-message-wrapper.type'
export * from './sqs-message-wrapper-params.interface'
export * from './sqs-message-handler.type'
