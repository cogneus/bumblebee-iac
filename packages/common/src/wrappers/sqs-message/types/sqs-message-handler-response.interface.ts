export interface SQSMessageHandlerResponse {
  success: boolean
  messageId: string
}
