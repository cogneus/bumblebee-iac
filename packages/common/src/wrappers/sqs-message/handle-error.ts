import { ErrorActionCollection, SQSMessageHandlerResponse } from './types'
import { Logger } from '../../logger'

export const handleError = (
  e: any,
  logger: Logger,
  messageId: string,
  errorActions?: ErrorActionCollection
): SQSMessageHandlerResponse => {
  const action = errorActions?.[e.code] || 'unknown'
  if (action !== 'ignore') {
    logger.warn(`Process failed, ${e.message}`)
    logger.error(e)
  }
  return {
    messageId,
    success: ['ignore', 'no-retry'].includes(action),
  }
}
