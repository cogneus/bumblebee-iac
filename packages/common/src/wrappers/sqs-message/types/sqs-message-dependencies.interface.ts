import { ServiceContext } from '../../../types'

export interface SQSMessageDependencies<TMessage> {
  message: TMessage
  messageId: string
  serviceContext: ServiceContext
}
