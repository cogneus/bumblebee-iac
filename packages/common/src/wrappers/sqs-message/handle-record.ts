import { SQSMessageHandlerParams, SQSMessageHandlerResponse } from './types'
import { applyWrapperClients, createWrapperContext } from '../utils'
import { handleError } from './handle-error'
import { destroyClients } from '../../clients'

export const handleRecord = async <TMessage>({
  record,
  context,
  handler,
  parse,
  getClientOptions,
  errorActions,
  ...contextParams
}: SQSMessageHandlerParams<TMessage>): Promise<SQSMessageHandlerResponse> => {
  const serviceContext = createWrapperContext({
    ...contextParams,
    context,
    logInfo: { record },
  })
  const { log } = serviceContext
  const { messageId } = record
  try {
    const message = parse(record)
    applyWrapperClients(getClientOptions(message, record), serviceContext)
    const result = await handler({
      serviceContext,
      message,
      messageId,
    })
    log.debug('Process complete', result)
    return result
  } catch (ex) {
    return handleError(ex, log, messageId, errorActions)
  } finally {
    destroyClients(serviceContext.clients)
  }
}
