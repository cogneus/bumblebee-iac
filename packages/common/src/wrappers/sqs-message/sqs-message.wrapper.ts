import 'source-map-support/register'
import type { SQSBatchItemFailure } from 'aws-lambda'
import { ServiceContext } from '../../types'
import { SQSMessageWrapper } from './types'
import { handleRecord } from './handle-record'

export const sqsMessageWrapper: SQSMessageWrapper =
  (params) => async (event, context: ServiceContext) => {
    const { enabled } = params
    if (!enabled) {
      return { batchItemFailures: [] }
    }
    const { Records } = event
    const results = await Promise.all(
      Records.map((record) => handleRecord({ ...params, record, context }))
    )
    const batchItemFailures: SQSBatchItemFailure[] = results
      .filter(({ success }) => !success)
      .map(({ messageId: itemIdentifier }) => ({
        itemIdentifier,
      }))

    return { batchItemFailures }
  }
