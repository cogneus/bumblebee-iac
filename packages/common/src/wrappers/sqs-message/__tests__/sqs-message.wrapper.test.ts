import { SQSEvent, SQSRecord } from 'aws-lambda'
import { sqsMessageWrapper } from '../sqs-message.wrapper'
import { AppException, StatusCodes } from '../../../exceptions'
import { ErrorActionCollection } from '../types'
import { ClientOptions } from '../../../clients'

const InvalidArgumentRetry: AppException = {
  name: 'invalid-argument-retry',
  code: 'invalid-argument-retry',
  message: 'test exception retry',
  status: StatusCodes.BadRequest,
}

const InvalidArgumentIgnore: AppException = {
  name: 'invalid-argument-ignore',
  code: 'invalid-argument-ignore',
  message: 'test exception ignore',
  status: StatusCodes.BadRequest,
}

const errors: {
  error: AppException | Error
  messageId: any
}[] = [
  {
    error: InvalidArgumentRetry,
    messageId: 1,
  },
  {
    error: InvalidArgumentIgnore,
    messageId: 2,
  },
  {
    error: new Error('Generic error'),
    messageId: 3,
  },
]

const clientOptions: ClientOptions = {
  db: {
    maxAttempts: 3,
    region: 'eu-west-1',
    requestTimeout: 3000,
    tables: {
      test: {
        name: 'test-table-name',
        indices: {
          testIndex: { name: 'test-index-name' },
        },
        itemLimits: { defaultLimit: 20, maxLimit: 100 },
      },
    },
  },
}

export const errorActions: ErrorActionCollection = {
  [InvalidArgumentRetry.code]: 'no-retry',
  [InvalidArgumentIgnore.code]: 'ignore',
}

const baseRecord: SQSRecord = {
  attributes: {
    ApproximateFirstReceiveTimestamp: '',
    ApproximateReceiveCount: '',
    SenderId: '',
    SentTimestamp: '',
  },
  awsRegion: 'any',
  body: '',
  eventSource: '',
  eventSourceARN: '',
  md5OfBody: '',
  messageAttributes: {},
  messageId: '',
  receiptHandle: '',
}

const messages: {
  event: SQSEvent
  response: {
    success: boolean
  }
  enabled: boolean
}[] = [
  {
    response: { success: true },
    event: { Records: [{ ...baseRecord, messageId: '1' }] },
    enabled: true,
  },
  {
    response: { success: true },
    event: {
      Records: [
        { ...baseRecord, messageId: '1' },
        { ...baseRecord, messageId: '2' },
      ],
    },
    enabled: true,
  },
  {
    response: { success: false },
    event: { Records: [{ ...baseRecord, messageId: '1' }] },
    enabled: true,
  },
  {
    response: { success: false },
    event: {
      Records: [
        { ...baseRecord, messageId: '1' },
        { ...baseRecord, messageId: '2' },
      ],
    },
    enabled: true,
  },
  {
    response: { success: false },
    event: {
      Records: [
        { ...baseRecord, messageId: '1' },
        { ...baseRecord, messageId: '2' },
      ],
    },
    enabled: false,
  },
]

describe('sqs message wrapper tests', () => {
  describe('should handle response, returns', () => {
    it.each(messages)('results matching snapshot', async ({ event, response, enabled }) => {
      jest.resetAllMocks()
      const wrapper = sqsMessageWrapper({
        handler: async ({ messageId }) => ({
          ...response,
          messageId,
        }),
        parse: ({ body }) => ({
          ...response,
          body,
        }),
        region: 'eu-west-1',
        env: 'dev',
        logName: 'wrapper-test',
        serviceName: 'wrapper-test',
        getClientOptions: () => clientOptions,
        enabled,
        errorActions,
      })
      const cb: any = null
      const result = await wrapper(event, {} as any, cb)
      expect(result).toMatchSnapshot()
    })
  })

  describe('should handle errors, returns', () => {
    it.each(errors)('results matching snapshot', async ({ error, messageId }) => {
      jest.resetAllMocks()
      const wrapper = sqsMessageWrapper({
        handler: async () => {
          throw error
        },
        parse: ({ body }) => ({
          body,
        }),
        region: 'eu-west-1',
        env: 'dev',
        logName: 'wrapper-test',
        serviceName: 'wrapper-test',
        getClientOptions: () => clientOptions,
        enabled: true,
        errorActions,
      })
      const cb: any = null
      const result = await wrapper(
        {
          Records: [
            {
              ...baseRecord,
              messageId,
            },
          ],
        },
        {} as any,
        cb
      )
      expect(result).toMatchSnapshot()
    })
  })
})
