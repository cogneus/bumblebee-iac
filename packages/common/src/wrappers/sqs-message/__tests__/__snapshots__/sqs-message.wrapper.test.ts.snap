// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sqs message wrapper tests should handle errors, returns results matching snapshot 1`] = `
{
  "batchItemFailures": [],
}
`;

exports[`sqs message wrapper tests should handle errors, returns results matching snapshot 2`] = `
{
  "batchItemFailures": [],
}
`;

exports[`sqs message wrapper tests should handle errors, returns results matching snapshot 3`] = `
{
  "batchItemFailures": [
    {
      "itemIdentifier": 3,
    },
  ],
}
`;

exports[`sqs message wrapper tests should handle response, returns results matching snapshot 1`] = `
{
  "batchItemFailures": [],
}
`;

exports[`sqs message wrapper tests should handle response, returns results matching snapshot 2`] = `
{
  "batchItemFailures": [],
}
`;

exports[`sqs message wrapper tests should handle response, returns results matching snapshot 3`] = `
{
  "batchItemFailures": [
    {
      "itemIdentifier": "1",
    },
  ],
}
`;

exports[`sqs message wrapper tests should handle response, returns results matching snapshot 4`] = `
{
  "batchItemFailures": [
    {
      "itemIdentifier": "1",
    },
    {
      "itemIdentifier": "2",
    },
  ],
}
`;

exports[`sqs message wrapper tests should handle response, returns results matching snapshot 5`] = `
{
  "batchItemFailures": [],
}
`;
